### PR TITLE
Fixes crash caused by invisible semantics children

### LINKED ISSUE #13326: Child with id 36 is invisible and should not be added to tree.
## Exception
```
══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
The following assertion was thrown during a scheduler callback:
Child with id 36 is invisible and should not be added to tree.
'package:flutter/src/semantics/semantics.dart': Failed assertion: line 370 pos 16:
'!child.isInvisible'

Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
https://github.com/flutter/flutter/issues/new

When the exception was thrown, this was the stack:
════════════════════════════════════════════════════════════════════════════════════════════════════
```

## flutter analyze
```
Analyzing /Users/hao/Documents/flutter_proj/shop...
No issues found!
Ran in 5.8s
```

## Flutter Doctor
```
[✓] Flutter (on Mac OS X 10.13.1 17B1003, locale zh-Hans-CN, channel alpha)
    • Flutter at /Users/hao/Documents/flutter
    • Framework revision d957c8f040 (3 days ago), 2017-11-30 13:29:59 -0800
    • Engine revision 77d8acb9be
    • Tools Dart version 1.25.0-dev.11.0
    • Engine Dart version 2.0.0-dev.9.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/hao/Library/Android/sdk
    • Unable to locate Android NDK.

    • Unable to locate compiler in Android NDK.

    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 9.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.1, Build version 9B55
    • ios-deploy 1.9.2
    • CocoaPods version 1.3.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Connected devices
    • iPhone 6s • AB7E1A1A-2F6B-469E-BA9F-BFAFA8C9A288 • ios • iOS 11.1 (simulator)

```

### What does this exception mean?

### PR DESCRIPTION
**Pull request by goderbauer**
**tl;dr:** A `RenderObject` can only be an effective semantics boundary if it actually owned a `SemanticsNode` in the previous tree generation.

When using the `BlockSemantics` widget it is possible to introduce `RenderObject`s that are configured to be a semantics boundary, but because their semantics are blocked by `BlockSemantics` they will not (immediately) end up owning a `SemanticsNode`. When now a descendant of such a node-less semantics boundary marks itself as needing a semantics update we walk up the tree until we find the closest semantics boundary (which is our node-less `RenderObject`). We now incorrectly assume that this semantics boundary has a valid `SemanticsNode` and only regenerate the semantics subtree below this node. However, because the identified semantics boundary doesn't actually own a valid `SemanticsNode` asserts are throwing (e.g. `Child with id xx is invisible and should not be added to tree.`).

To fix this problem, we can just abort the walk if we reach a semantics boundary without a semantics node because (for now) we know that the semantics information of this branch will not make it into the final semantics tree.
If the semantics block is ever removed, the current algorithm re-generates the semantics for the entire branch and the semantics will be up-to-date then despite the abort. I've added a test to verify this to make sure it continues to work even when we change the algorithm.

Fixes https://github.com/flutter/flutter/issues/13326.
/cc @gavindoughtie FYI

### CODE REVIEW

**Comment by Hixie** [comment]
Can we notice that it's a semantics boundary without a semantics node and assume that means that any semantics we create will be thrown away?

**Comment by Hixie** [comment]
LGTM

**Comment by goderbauer** [comment]
I tweaked this slightly to avoid doing extra work for `RenderObject`s whose semantics are currently blocked and extended the test to ensure that the semantics tree is correct when the block is removed.

I also made `BlockSemantics` configureable following the pattern also used for `IgnorePointer` and `ExcludeSemantics`.

**Comment by Hixie** [comment]
LGTM
